### PR TITLE
[ML] Restore analytics state if available

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
This commit restores the model state if available in data
frame analytics jobs.

In addition, this changes the start API so that a stopped job
can be restarted. As we now store the progress in the state index
when the task is stopped, we can use it to determine what state
the job was in when it got stopped.

### CODE REVIEW

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java]: Do we need to validate the mappings can be merged if we are already analyzing and have already reindexed all the data?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java]: seems like a `debug` level to me?

Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java]: Good catch, that was a leftover. We don't really need it at all.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartDataFrameAnalyticsAction.java
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartDataFrameAnalyticsAction.java]: Good catch. This is optional.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartDataFrameAnalyticsAction.java
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartDataFrameAnalyticsAction.java]: I've changed the constructor used by the parser to pass an empty list when `progress_on_start` is missing. The main constructor is now calling `Collections.unmodifiableList` which throws if the list is `null`.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartDataFrameAnalyticsAction.java
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartDataFrameAnalyticsAction.java]: On version 7.5.0 onwards it will never be null. I'm addressing the null issues in total.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartDataFrameAnalyticsAction.java
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartDataFrameAnalyticsAction.java]: Not yet. I'll add them right after merging this one in to save the trouble with flipping the versions here.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java
Comment (dimitris-athanasiou) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDataFrameAnalyticsAction.java]: If we are in `resuming_reindexing` we have not necessarily reindexed all the data. In general, I'd rather we perform a redundant check here and keep the path the same rather than skipping it and trying to reason about whether there are some cases that the mappings check would be necessary.

**Review by benwtrent — APPROVED** [review]

**Comment by dimitris-athanasiou** [comment]
run elasticsearch-ci/2

**Comment by dimitris-athanasiou** [comment]
run elasticsearch-ci/bwc

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

**Pull request merged**
